pkg/lib/import-json.ts

```diff
@@ -0,0 +1,182 @@
+           }
+       }
+
+   This way, TypeScript will check that the returned values are indeed
+   valid for their type declaration. You can't get that wrong. What is
+   not checked is that you use the right field names when accessing
+   input JsonObjects. But we could write a linter function for that.
+
+   More examples can be found in "pkg/shell/manifests.ts".
+ */
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string): never {
+    throw new ValidationError(`JSON validation error for ${validation_path.join("")}: ${msg}`);
+}
+
+export function import_string(val: JsonValue): string {
+    if (typeof val == "string")
+        return val;
+    validation_error(`Not a string: ${JSON.stringify(val)}`);
+}
+
+export function import_number(val: JsonValue): number {
+    if (typeof val == "number")
+        return val;
+    validation_error(`Not a number: ${JSON.stringify(val)}`);
+}
+
+export function import_boolean(val: JsonValue): boolean {
+    if (typeof val == "boolean")
+        return val;
+    validation_error(`Not a boolean: ${JSON.stringify(val)}`);
+}
+
+function is_json_object(val: JsonValue): val is JsonObject {
+    return !!val && typeof val == "object" && !Array.isArray(val);
+}
+
+export function import_json_object(val: JsonValue): JsonObject {
+    if (is_json_object(val))
+        return val;
+    validation_error(`Not an object: ${JSON.stringify(val)}`);
+}
+
+function is_json_array(val: JsonValue): val is JsonValue[] {
+    return !!val && typeof val == "object" && Array.isArray(val);
+}
+
+export function import_json_array(val: JsonValue): JsonValue[] {
+    if (is_json_array(val))
+        return val;
+    validation_error(`Not an array: ${JSON.stringify(val)}`);
+}
+
+export function import_record<T>(val: JsonValue, importer: (val: JsonValue) => T): Record<string, T> {
+    const obj = import_json_object(val);
+    return Object.fromEntries(Object.entries(obj).map(
+        ([k, v]) => [k, with_validation_path(`.${k}`, () => importer(v))]));
+}
+
+export function import_array<T>(val: JsonValue, importer: (val: JsonValue) => T): Array<T> {
+    const arr = import_json_array(val);
+    return arr.map((elt, i) => with_validation_path(`[${i}]`, () => importer(elt)));
+}
+
+export function import_optional<T>(obj: JsonObject, field: string, importer: (val: JsonValue) => T): T | undefined {
+    if (field in obj)
+        return with_validation_path(`.${String(field)}`, () => importer(obj[field]));
+    return undefined;
+}
+
+export function import_mandatory<T>(obj: JsonObject, field: string, importer: (val: JsonValue) => T): T {
+    if (field in obj)
+        return with_validation_path(`.${String(field)}`, () => importer(obj[field]));
```

Comment:
It might be interesting to capture the error raised by `importer` and augment it.  Consider some document like `{x: 0, y: 0, z: 0}` and then we call `import_mandatory(obj, "x", import_string)` and get the error "Not a string: 0".

Reply:
Sorry.  I just found `with_validation_path()` and take this back.

But: not excellent how this is effectively implemented via global variable... do we have any other reasonable options there?

Reply:
The best way would indeed be to catch the error and augment it. I'll do that.

Reply:
Done.

Reply:
This looks a lot nicer than it was before and it was my main real gripe from the previous iteration.